# [X-Village] Lesson03 - Object and Function
speaker: Amber（吳思嬋）

# Object
---
**In Python, we do things with stuff.**

## What is an object ?
---

### Everything in  Python  is an object !!!

## Built-in Types Object
---
Numbers, Strings, Lists, Dictionarys, Tuple, Files, Sets, etc.

## Built-in Types Object example
---

In [ ]:
# Number
12 # integer
12.5 # float-point
-12.5 # negative

# String
'Hello Word'

# list
[1, 2, 3, 4]

In [ ]:
# variable a, object
a = 3
a = 'Hello World'

## The Dynamic Typing Interlude
---
In Python, types are determined automatically at runtime.

## Variables, Objects, and References

## Variables (i.e. name)
---

**Variables are created when assigned, can reference any type of object, and must be assigned before they are referenced.**
* Variable creation
* Variable types
* Variable use

In [ ]:
a = 3 # Assign a name to an object

# 1. Create an object to represent the value 3.
# 2. Create the variable a, if it does not yet exist. 
# 3. Link the variable a to the new object 3.

![](https://i.imgur.com/qaynkKr.png)

### Types Live with Objects, Not Variables

In [ ]:
a = 3 # It's an integer
a = 'spam' # Now it's a string
a = 1.23 # Now it's a floating point

## Shared References (Shared Object)
---
![](https://i.imgur.com/ywCbsuH.png)

![](https://i.imgur.com/fsAyUyM.png)

![](https://i.imgur.com/Y27ye4P.png)

## Shared References and In-Place Changes
---
* immutable objects
    * int, float, bool, str, etc.
* mutable objects
    * list, set, dict, etc.

<img src="https://i.imgur.com/HWvxQCS.png" width=450 height=180 />
<img src="https://i.imgur.com/xGOu8Qs.png" width=450 height=180 />

## Objects are Garbage-Collected
---
Whenever a name is assigned to a new object, the space held by the prior object is reclaimed if it is not referenced by any other name or object.

In [ ]:
a = 3 # It's an integer
a = 'spam' # Now it's a string

In [ ]:
x = 42
x = 'shrubbery' # Reclaim 42 now (unless referenced elsewhere)
x = 3.1415 # Reclaim 'shrubbery' now
x = [1, 2, 3]  # Reclaim 3.1415 now

### Exercise1
經過這三個statements, print(a)的結果為何？

In [ ]:
# exercise1
a = "Hello"
b = a
b = "World"
print(a)

In [2]:
# answer
a = "Hello"
b = a
b = "World"
print(a)

Hello


### Exercise2
經過這三個statements, print(a)的結果為何？

In [ ]:
# exercise
a = ["Hello"]
b = a
b[0] = "World"
print(a)

In [3]:
# answer
a = ["Hello"]
b = a
b[0] = "World"
print(a)

['World']


# Function
---
1. **Function Basics**
2. Scopes
3. Arguments
4. Recursive Functions

## Coding Function: def Statements (1)
* keyword: def
* fucntion name, argument, return statement

![](https://i.imgur.com/cCia0rA.png)

In [4]:
# example1
def f():
    # function body
    print("Hello World")

f() # function call

Hello World


In [5]:
# example2
def f(n):
    # function body
    print(n)

x = "Hello World"
f(x) # function call

Hello World


In [1]:
# example 3
def g(n):
    # function body
    tmp = 1
    for i in range(1, n+1 ,1):
        tmp = tmp * i
    return tmp

x = g(5) # function call, 5*4*3*2*1
print(x)

120


In [7]:
# example 4
# Computes the total cost, including 3% sales tax
# on number items at a cost of price each
def total_cost(number, price):
    # function body
    tax_rate = 0.03 # 3% sales tax
    cost = number * price
    total = cost + (cost*tax_rate)
    return total

n = 5
p = 30
bill = total_cost(n, p)
print('Bill: ',bill)

Bill:  154.5


### enclosing function
---
![](https://i.imgur.com/XCS7yHE.png)

### Question
如果撰寫程式時，將 return object 的 object 省略，在呼叫 function 的階段會發生什麼事？

In [ ]:
# return statement without object
def f():
    x = 5
    return

y = f() # function call
print(y)

In [8]:
# answer
def f():
    x = 5
    return

y = f() # function call
print(y)

None


In [9]:
# without return statement
def f():
    x=5

y = f()
print(y) 

None


## Coding Function: def Statements (2)
---
* **def statement** 是 executable code
* keyword **def** 會建立 function object, 將 fucntion object assign 給 function name
* keywood **return** 會回傳 object 給 caller
* 傳遞 Arguments 的方式，使用 assignment
* Arguments, return objects, 和 variables 都不需要事先宣告型別

## Why use function
* **Maximizing code reuse**
* Procedural decompsition

In [10]:
# 公式：體重(公斤) / 身高^2(公尺^2)
# 過輕：BMI ＜ 18.5; 正常：18.5≦ BMI＜24; 過重：24≦BMI
def bmi(height, weight):
    calculate = weight / (height*height)
    if calculate < 18.5:
        result = '過輕'
        return result
    elif 18.5 < calculate and calculate < 24:
        result = '正常'
        return result
    else:
        result='過重'
        return result
    
h = 1.63 #公尺
w = 50 #公斤
r = bmi(h, w)
print("result1: ", r) # 正常

h = 1.80 #公尺
w = 88 #公斤
r = bmi(h, w)
print("result2: ", r) # 過重

result1:  正常
result2:  過重


In [12]:
height = 1.63 #公尺
weight = 50 #公斤
calculate = weight / (height*height)
if calculate < 18.5:
    result = '過輕'
elif 18.5 < calculate and calculate < 24:
    result = '正常'
else:
    result='過重'
print("result1: ", result)
    
height = 1.80 #公尺
weight = 88 #公斤
calculate = weight / (height*height)
if calculate < 18.5:
    result = '過輕'
elif 18.5 < calculate and calculate < 24:
    result = '正常'
else:
    result='過重'
print("result2: ", result)

result1:  正常
result2:  過重


## Why use function
* Maximizing code reuse
* **Procedural decompsition**

In [ ]:
# Procedural decopsition example
# pizza-making robot
def mixing_the_dough() #和麵
def rolling_it_out() #桿麵
def adding_toppings() #放料
def baking_it() #烤

### In general, functions are about procedure
* how to do something, rather than what you’re doing it to ! 

### Exercise
---
寫一個function，用途計算數字的平方根。
* example
    * 4的平方根為2
    * 9的平方根為3
    * 2的平方根為1.4142....

In [13]:
import math
x = math.sqrt(4)
print(x)

y = math.sqrt(2)
print(y)

2.0
1.4142135623730951


### Do not reinvent the wheel !! 

### Exercise1: MxM_multiplication (5min) - 列入課堂成績
---
請定義一個function f，參數n。
呼叫function時，傳入整數m，function 會印出 m x m 的乘法表。

* 提示：print 使用 end='' 可以不印出換行符號'\n'
* example:
    * f(9)
    * [output](https://i.imgur.com/0w5ajUv.png)

### Exercise2: swap (5min) - 列入課堂成績
---
定義一個function f，可以將 a, b 的值互換後，需回傳值回來。
* example: 
    * a=1, b=2
    * (a, b) = f(a, b)
    * output: a=2, b=1
    
### Exercise3: judge_big_lower (5min) - 列入課堂成績
---
定義一個function f， 判斷a, b 大小關係， 回傳字串結果, a>b, a=b 或 a<b 
* example: 
    * a=1, b=2
    * result = f(a,b)
    * output: a<b

# Function
---
1. Function Basics
2. **Scopes**
3. Arguments
4. Recursive Functions

## Nampespace
---
![](https://i.imgur.com/RRDpUAZ.png)

## Python Scope Basics 
---
1. assign name inside **def**： local to that function
2. assign name in **enclosing def**, nonlocal to nested functions
3. assign name  outside all **defs**, global to the entire file.


In [ ]:
x = 99 # Global (module) scope x 
def f():
    x = 88 # Local (function) scope x: a different variable

## The LEGB Rule
---
**當我們使用variable, python 在scope中搜尋variable的順序為 L, E, G 最後是 B**
* **L:** local scope
* **E:** local scopes of any enclosing defs
* **G:** global scope 
* **B:** built-in scope

## The LEGB Rule
---
<img src="https://i.imgur.com/5am8Cm4.png" width=450 height=180 />

## Scope example
---
* Global names: X, f
* Local names: Y, Z


In [ ]:
# Global scope
x = 99  # X and f assigned in module: global
def f(y): # Y and Z assigned in function: locals
    # Local scope
    z = x + y # X is a global
    return z

f(1) # function1 in module: result=100

## The Built-in Scope
---
* 在 module *builtins* 中定義的name, 都屬於  built-in scope

In [17]:
import builtins
dir(builtins)

['ArithmeticError',
 'AssertionError',
 'AttributeError',
 'BaseException',
 'BlockingIOError',
 'BrokenPipeError',
 'BufferError',
 'BytesWarning',
 'ChildProcessError',
 'ConnectionAbortedError',
 'ConnectionError',
 'ConnectionRefusedError',
 'ConnectionResetError',
 'DeprecationWarning',
 'EOFError',
 'Ellipsis',
 'EnvironmentError',
 'Exception',
 'False',
 'FileExistsError',
 'FileNotFoundError',
 'FloatingPointError',
 'FutureWarning',
 'GeneratorExit',
 'IOError',
 'ImportError',
 'ImportWarning',
 'IndentationError',
 'IndexError',
 'InterruptedError',
 'IsADirectoryError',
 'KeyError',
 'KeyboardInterrupt',
 'LookupError',
 'MemoryError',
 'ModuleNotFoundError',
 'NameError',
 'None',
 'NotADirectoryError',
 'NotImplemented',
 'NotImplementedError',
 'OSError',
 'OverflowError',
 'PendingDeprecationWarning',
 'PermissionError',
 'ProcessLookupError',
 'RecursionError',
 'ReferenceError',
 'ResourceWarning',
 'RuntimeError',
 'RuntimeWarning',
 'StopAsyncIteration',
 'StopIter

## Enclosing function scope
---
* **E:** local scopes of any enclosing defs

### Enclosing function example
---
* **Global names: W, outer**
* **Enclosing local names: X, inner**
* **Local names: Z, Y**

In [ ]:
# Global scope
w = 21  # W and outer assigned in module: global
def outer(x):# X and inner assigned in enclosing function: Enclosing def locals
    def inner(y): # Z, Y assigned in function: local
        z = w + x + y  # W is a global (21+88+1)
        print(z)
    inner(1)
outer(88)

### Exercise

In [ ]:
# exercise1
x = 'Spam'
def func():
    print(x)
    
func()

In [18]:
# answer
x = 'Spam'
def func():
    print(x)
    
func()

Spam


In [ ]:
# exercise2
x = 'Spam'
def func():
    x = 'HI!'
    
func()
print(x)

In [19]:
# answer
x = 'Spam'
def func():
    x = 'HI!'
    
func()
print(x)

Spam


In [ ]:
# exercise3
x = 'World'
def func():
    x = 'Hello'
    print(x)
    
func()
print(x)

In [20]:
# answer
x = 'World'
def func():
    x = 'Hello'
    print(x)
    
func()
print(x)

Hello
World


In [ ]:
# exercise 4
w = 22
def outer():
    w = 23
    def inner():
        w = 24
        print('first: ', w)  # first print
    inner()
    print('second: ', w) # second print
    
outer()
print('third: ', w) # third print

In [21]:
# answer
w = 22
def outer():
    w = 23
    def inner():
        w = 24
        print('first: ', w)  # first print
    inner()
    print('second: ', w) # second print
    
outer()
print('third: ', w) # third print

first:  24
second:  23
third:  22


## The global Statement
---
* keyword: **global**

In [22]:
# example
x = 88 # Global x
def f(): 
    global x
    x = 99 # Global x: outside def

f()
print(x)

99


## Factory Function: Closures
---
The function object remembers values in enclosing scopes regardless of whether those scopes are still present in memory.


In [23]:
# This defines an outer function that simply generates and returns a nested function without calling it.
# maker makes action, but returns action without running it.
def maker(n):
    def action(x): # Generate and return action
        return x * n # action retains N from enclosing scope return action
    return action

f = maker(2) # Pass 2 to argument N
print(f)
print(f(3)) # Pass 3 to X, N remembers 2: 3*2=6
print(f(4)) # Pass 4 to X, N remembers 2: 4*2=8

<function maker.<locals>.action at 0x1096b1ae8>
6
8


In [24]:
def maker(n):
    def action(x): # Generate and return action
        return x * n # action retains N from enclosing scope return action
    return action

f = maker(2) # Pass 2 to argument N
g = maker(3) # Pass 3 to argument N

print(f)
print(f(4)) # Pass 4 to X, N remembers 2: 2*4=4

print(g)
print(g(4)) # Pass 4 to X, N remembers 3: 3*4=6

<function maker.<locals>.action at 0x1096b1c80>
8
<function maker.<locals>.action at 0x1096b1ae8>
12


In [25]:
# advanced technique
# lambda arg1, arg2, ....: expression
def maker(n):
     return lambda x: x * n
    
h = maker(3)
print(h(4))

12


### Exercise1: X^N (10min) - 列入課堂成績
---
請試著用  Factory Function 的方法， 計算 N 的 X 次方。
* example:
    * **f = maker(9)** : N=9
    * **f(5)**  : X=5
    * **print(f(5))**: 59049 (為9的5次方)

### Exercise2: carbon_foot (10min) - 列入課堂成績
---
請試著用  Factory Function 的方法， 計算交通使用 公車, 汽車, 機車 的碳足跡排放量。

* 碳足跡計算方式
    * rate * km
    * rate(kg/km): 公車:0.03, 汽車:0.24, 機車:0.06

* example
    * way = carbon_foot('car') 
    * way(100): 0.24 * 100 = 24.0 (kg/km)

# Function
---
1. Function Basics
2. Scopes
3. **Arguments**
4. Recursive Functions

Review: Object
---
**Everything in python is an object**

<img src="https://i.imgur.com/qaynkKr.png" width=450 height=180 />

In [28]:
a = 3
print(id(a), id(3))

4412807904 4412807904


## Review: immutable object
---
![](https://i.imgur.com/YCRrRgE.png)

## Review: mutable object
---
<img src="https://i.imgur.com/HWvxQCS.png" width=450 height=180 />
<img src="https://i.imgur.com/xGOu8Qs.png" width=450 height=180 />

##  Argument-Passing Basics (1)
---
* **aliasing**
    * assignment to an argument name inside a function (e.g., a=99) does not change a variable like b in the scope of the function call.

In [29]:
def f(a): # a is assigned to (references) the passed object
    a = 99 # Changes local variable a only
b = 88 
f(b) # a and b both reference same 88 initially
print(b) # b is not changed

88


In [30]:
def changer(b): # Arguments assigned references to objects
    print('before_b:',b)
    b[0] = 'spam' # Changes shared object in place
    print('after_b:',b)
    print('b_id:',id(b))
    
l = [1, 2] 
changer(l)  # Pass mutable object
print('L:', l) # l is different
print('L_id:', id(l))

before_b: [1, 2]
after_b: ['spam', 2]
b_id: 4452870792
L: ['spam', 2]
L_id: 4452870792


##  Argument-Passing Basics (2)
---
**Avoiding Mutable Argument Changes**: 
If we don’t want in-place changes within functions to impact objects we pass to them, though, we can simply make explicit copies of mutable objects,

In [31]:
def changer(b): # Arguments assigned references to objects
    b[0] = 'spam' # Changes shared object in place
    
L = [1, 2]
copyL=L[:] 
print('before: L:', L, '; L_id:', id(L))
print('before: copyL:', copyL, '; copyL_id:', id(copyL))
changer(copyL)
print('after: L:', L, '; L_id:', id(L))
print('after: copyL:', copyL, '; copyL_id:', id(copyL))

before: L: [1, 2] ; L_id: 4452815432
before: copyL: [1, 2] ; copyL_id: 4452712584
after: L: [1, 2] ; L_id: 4452815432
after: copyL: ['spam', 2] ; copyL_id: 4452712584


## Argument-Passing Basics (3)
---
* Arguments are passed by automatically assigning objects to local variable names

* Assigning to argument names inside a function does not affect the caller

* Changing a mutable object argument in a function may impact the caller.


## Argument Matching Basics (1)
---
* **Positionals:** matched from left to right

In [32]:
# positionals example
def f(x, y, z):
    print(x, y, z)
f(1, 2, 3)

1 2 3


## Argument Matching Basics (2)
---
* **Keywords:** matched by argument name

In [33]:
# keywords example
def f(x, y, z):
    print(x, y, z)
    
f(z=4, y=5, x=6)

6 5 4


## Argument Matching Basics (3)
---
* **Defaults:** specify values for optional arguments that aren’t passed

In [34]:
# defaults and postional
def f(x, y, z=3):
    print(x, y, z)
f(4, 5)
f(4, 5, 6)

4 5 3
4 5 6


## Argument Matching Basics
---
* **Varargs collecting:** collect arbitrarily many positional or keyword arguments


In [35]:
# varargs collecting, one * for tuple
def f1(*args):
    print(args)
f1(1, 2, 3, 4, 5, 6, 8, 'hello')
f1('how', 'do', 'you', 'do')

def f2(x, y, *args):
    print(x,y)
    print(args)
f2(1,2,3,4,5,6,8,'hello')

(1, 2, 3, 4, 5, 6, 8, 'hello')
('how', 'do', 'you', 'do')
1 2
(3, 4, 5, 6, 8, 'hello')


In [36]:
# varargs collecting, two * for dictionary
def f(**kwargs):
    print(kwargs)
f(key1=1, key2=2, key3=3)

{'key1': 1, 'key2': 2, 'key3': 3}


In [37]:
# mix varargs collecting
def mix(x, y, *args, **kwargs):
    print(x,y)
    print(args)
    print(kwargs)
mix(1, 2, 3, 4, 5, 6, 8, 'hello', key1=1, key2=2, key3=3)

1 2
(3, 4, 5, 6, 8, 'hello')
{'key1': 1, 'key2': 2, 'key3': 3}


## Argument Matching Basics
---
* **Varargs unpacking:** pass arbitrarily many positional or keyword arguments

In [39]:
# Varargs unpacking 
def f(x, y, z):
    print(x, y, z)
    
args=(1, 2, 3)
f(*args)

kwargs = {'x':1, 'y':2, 'z':3 }
f(**kwargs)

1 2 3
1 2 3


### Exercise: Argument_Matching 

In [ ]:
# exercise
# What is the output of the following code？
def f1(a, b=4, c=5):
    print(a, b, c)
f1(1, 2)

def f2(a, b, c=5): 
    print(a, b, c)
f2(1, c=3, b=2)

def f3(a, *args):
    print(a, args)
f3(1, 2, 3)

def f4(a, **kargs): 
    print(a, kargs)
f4(a=1, c=3, b=2)

def f5(a, b, c=3, d=4): 
    print(a, b, c, d)
f5(1, *(5, 6))

##  Function Design Concepts
---
* use arguments for inputs and return for outputs
* use global variables only when truly necessary
* don’t change mutable arguments unless the caller expects it.
* each function should have a single, unified purpose.

## Function Objects
---
* Python functions
    * are full-blown objects, stored in pieces of memory all their own. 
    * can be freely passed around a program and called indirectly.
    * support operations that have little to do with calls at all—attribute storage and annotation.

In [40]:
# Function objects example: assigned to name
def f(msg):
    print(msg)

f('hello word!')
x = f
x('hello word!')

hello word!
hello word!


In [42]:
# Function objects example: pass to other function
def f(msg):
    print(msg)
    
def g(f_obj, arg): 
    f_obj(arg)

g(f, 'Hello world!') # Pass the function to another function

Hello world!


In [43]:
# Function objects example: embedded in data structures
def f(msg):
    print(msg)
    
L =[(f, 'Hello'), (f, 'World!!')]

for (f_obj, arg) in L:
    f_obj(arg) # Call functions embedded in containers

Hello
World!!


In [2]:
# Function objects example: return
def f(msg):
    def g(msg2):
        print(msg+':'+msg2)
    return g

t=f('hello')
t('world!')

hello:world!


# Function
---
1. Function Basics
2. Scopes
3. Arguments
4. **Recursive Functions**

## Recursive Functions
---
**Functions that call themselves either directly or indirectly in order to loop**

In [44]:
# sum for 1+2+3+...+10
def f(n):
    print('n:',n) # Trace recursive levels
    if n==1:
        return 1 
    else:
        return f(n-1)+n

print('answer: ',f(10))

# f(10)= f(9)+10 -> next level 
# f(9) = f(8)+9 -> next level 
# f(8) = f(7)+8 -> next level 
# ...
# f(3)=f(2)+3 -> next level 
# f(2)=f(1)+2 -> next level 
# f(1) return 1 -> next level 

# f(2)=f(1)+ 2 = 1 + 2 -> previous level, f(1) is 1 now
# f(3)=f(2)+ 3 = 3 + 3 -> previous level, f(2) is 3 now
#...

n: 10
n: 9
n: 8
n: 7
n: 6
n: 5
n: 4
n: 3
n: 2
n: 1
answer:  55


### exercise1: recursive_1 ( 10 min ) - 列入課堂成績
---
仿照上一個例子，請試著寫數字連乘的遞迴函式，計算出f(10)等於多少。
* f(5) = 1 x 2 x 3 x 4 x 5 = 120
* f(6)= 1 x 2 x 3 x 4 x 5 x 6 = 720

### exercise2: recursive_2 ( 10 min ) - 列入課堂成績
---
請試著寫數字連乘的遞迴函式，計算出費氏(Fibonacci)數列 F(20)，公式如下：
* F(0) = 0
* F(1) = 1
* F(n) = F(n-1) + F(n-2), when n≥2

![](https://i.imgur.com/jQjrefa.png)

## Loop Statements Versus Recursion
---

In [47]:
# recursion
def f(n):
    if n==1:
        return 1 
    else:
        return f(n-1) + n
print('receusion: ', f(10))

# loop
total = 0
for i in range(11): # 0, 1, ,2 ,3 ,4 , ..., 10
    total = total + i
print('loop', total)

receusion:  55
loop 55


# 課堂練習 - [河內塔問題](https://zh.wikipedia.org/wiki/%E6%B1%89%E8%AF%BA%E5%A1%94)
---
#### 有三根杆子A，B，C。A杆上有N個(N>1)穿孔圓盤，盤的尺寸由下到上依次變小。要求按下列規則將所有圓盤移至C杆：

* 每次只能移動一個圓盤
* 大盤不能疊在小盤上面。
* 可將圓盤臨時置於B杆，也可將從A杆移出的圓盤重新移回A杆，但都必須遵循上述兩條規則。

* ** Question: 5個圓盤，總共要移動多少次，移動的順序為何？ **

![](https://i.imgur.com/zbKUzwu.gif)